### PR TITLE
Лысов Иван. Задача 1. Вариант 2. Технология SEQ. Умножение плотных матриц. Элементы типа double. Блочная схема, алгоритм Фокса.

### PR DESCRIPTION
**Pull request by Ivan-Lysov**
### 1. **PreProcessingImpl()**
**Description**: Prepares the data for executing the algorithm. Initializes matrices \( A \), \( B \), and \( C \), and copies input data into the corresponding vectors.
**Main Steps**:
- Extracts the matrix size \( N \) and block size `block_size` from the input data.
- Initializes vectors \( A \), \( B \), and \( C \) - result matrix.
- Copies data from input arrays into vectors \( A \) and \( B \).

---

### 2. **ValidationImpl()**
**Description**: Validates the correctness of the input data.

**Main Steps**:
- Checks that the number of input and output data matches the expected values.
- Ensures that the sizes of matrices \( A \) and \( B \) are \( N \times N \).
- Verifies that the block size `block_size` is greater than 0.

---

### 3. **RunImpl()**
**Description**: Executes the Fox algorithm for matrix multiplication.

**Main Steps**:
- Splits matrices \( A \) and \( B \) into blocks of size `block_size`.
- Performs a cyclic shift of blocks in matrix \( A \).
- Multiplies blocks of matrices \( A \) and \( B \), accumulating the result in matrix \( C \).

---

### 4. **PostProcessingImpl()**
**Description**: Saves the result of matrix multiplication into the output data.

**Main Steps**:
- Copies data from vector \( C \) into the output array.

### CODE REVIEW

**Review by sozozzya — APPROVED** [review]
Please add tests with negative values of the matrix elements.

**Review by sozozzya — COMMENTED** [review]
Please add tests with negative values of the matrix elements.

File: tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+#include <memory>
+#include <random>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+#include "seq/lysov_i_matrix_multiplication_Fox_algorithm/include/ops_seq.hpp"
+namespace {
+void TrivialMatrixMultiplication(const std::vector<double> &matrix_a, const std::vector<double> &matrix_b,
+                                 std::vector<double> &result_matrix, size_t matrix_size);
+std::vector<double> GetRandomMatrix(size_t size);
```
Comment (ascannel) [inline tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp:15]: i suggest you to move this parts in header and use your own named namespace

File: tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+  task_data_sequential->inputs.emplace_back(reinterpret_cast<uint8_t *>(a.data()));
+  task_data_sequential->inputs.emplace_back(reinterpret_cast<uint8_t *>(b.data()));
+  task_data_sequential->inputs.emplace_back(reinterpret_cast<uint8_t *>(&block_size));
+  task_data_sequential->outputs.emplace_back(reinterpret_cast<uint8_t *>(c.data()));
+  task_data_sequential->inputs_count.emplace_back(n * n);
+  task_data_sequential->inputs_count.emplace_back(b.size());
+  task_data_sequential->inputs_count.emplace_back(1);
+  task_data_sequential->outputs_count.emplace_back(n * n);
+  lysov_i_matrix_multiplication_fox_algorithm_seq::TestTaskSequential matrix_multiplication(task_data_sequential);
+  ASSERT_EQ(matrix_multiplication.ValidationImpl(), false);
```
Comment (ascannel) [inline tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp:240]: you may just use assert_false

File: tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/src/ops_seq.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
+#include "seq/lysov_i_matrix_multiplication_Fox_algorithm/include/ops_seq.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+#include <vector>
+namespace {
```
Comment (ascannel) [inline tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/src/ops_seq.cpp:7]: again, use your own namespace

**Review by ascannel — CHANGES_REQUESTED** [review]

File: tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp (hunk 1 above)
Comment (Ivan-Lysov) [reply tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp]: fix, thank you

File: tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp (hunk 2 above)
Comment (Ivan-Lysov) [reply tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp]: fix, thank you

File: tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/src/ops_seq.cpp (hunk 3 above)
Comment (Ivan-Lysov) [reply tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/src/ops_seq.cpp]: fix, thank you

**Review by ascannel — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]

File: tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+#include "core/task/include/task.hpp"
+#include "seq/lysov_i_matrix_multiplication_Fox_algorithm/include/ops_seq.hpp"
+
+namespace lysov_i_matrix_multiplication_fox_algorithm_seq {
+std::vector<double> GetRandomMatrix(size_t size) {
+  std::vector<double> matrix(size * size);
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_real_distribution<double> dist(1e-3, 1e3);
```
Comment (allnes) [inline tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp:18]: please use negative values

File: tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+#include <cstdint>
+#include <memory>
+#include <random>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+#include "seq/lysov_i_matrix_multiplication_Fox_algorithm/include/ops_seq.hpp"
+
+namespace lysov_i_matrix_multiplication_fox_algorithm_seq {
+std::vector<double> GetRandomMatrix(size_t size) {
```
Comment (allnes) [inline tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp:14]: please parametrize range

**Review by allnes — APPROVED** [review]

File: tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp (hunk 4 above)
Comment (Ivan-Lysov) [reply tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp]: fix, thank you

File: tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp (hunk 5 above)
Comment (Ivan-Lysov) [reply tasks/seq/lysov_i_matrix_multiplication_Fox_algorithm/func_tests/main.cpp]: fix, thank you
